### ISSUE TITLE
[selectors] :link and :visited are not mutually exclusive in implementations

### ISSUE THREAD

**Issue by victoriasu** [w3c/csswg-drafts#2037]
According to the [spec](https://drafts.csswg.org/selectors-4/#link), :link and :visited should be mutually exclusive, however in:
Chrome, Firefox, and Safari
- Style rules for :link and :visited are being applied to the same element (a visited link)
Edge
- All links are using the :link style and not the :visited style

Tested using:
https://jsfiddle.net/victoriaytsu/ujz85sor/

**Comment by lilles** [comment]
This is for privacy reasons. See: https://developer.mozilla.org/en-US/docs/Web/CSS/Privacy_and_the_:visited_selector

Your spec link says:

"Since it is possible for style sheet authors to abuse the :link and :visited pseudo-classes to determine which sites a user has visited without the user’s consent, UAs may treat all links as unvisited links or implement other measures to preserve the user’s privacy while rendering visited and unvisited links differently."

So Edge is doing the former thing and Chrome/Firefox/Safari are doing the "other measures"?

**Comment by emilio** [comment]
Yeah, querySelector and such never match `:visited` on Gecko. Similarly `getComputedStyle` doesn't return `:visited` styles and such.

**Comment by dbaron** [comment]
For a single selector matching operation, I think they still are mutually exclusive, though.  At least based on [what I originally described](https://dbaron.org/mozilla/visited-privacy).

**Comment by FremyCompany** [comment]
I think the difference in Edge is just an invalidation issue for ":link" because we are supposed to behave like other browsers here. https://jsfiddle.net/gck56f1h/ shows we still apply an outline but with wrong color once we visit the link, it doesn't make much sense to me.

**Comment by FremyCompany** [comment]
Though actually loading the demo in other browsers, I am not sure I understand their behavior either. I think the problem is that you cannot remove all ":link" styles when the link is visited because otherwise this might leak somehow, so only a couple of properties will match using ":visited" and other properties will match using ":link" and this creates a weird state.

Maybe saying ":link" always matches would help solve the craziness but there might be a few websites broken by specificity of ":link" overriding their ":visited" styles.

**Comment by FremyCompany** [comment]
Looking at the test case a bit more, here is what I understand the difference between Edge and other browsers to be:

Edge:
- Apply all :link rules and, for allowed properties, also apply all :visited rules (at the same time)

Other browsers:
- Apply all rules that do not have :visited (including :link rules)
- Apply all rules that have :visited or do have neither :link nor :visited

The difference is the cascade order. Effectively, a :link rule can override a :visited rule in Edge if its specificity is higher. In Firefox and Chrome, it cannot because this is done as part of a clean second pass.

https://jsfiddle.net/ngtby4te/ shows this (lime outline but yellow background in Edge)

Perf-wise, our behavior seems better. Not sure this is something we want to fix, I am yet to see any issue appear because of that.

**Comment by emilio** [comment]
> Apply all :link rules and, for allowed properties, also apply all :visited rules (at the same time).

As I understand it, assuming you only apply `:visited` rules to actually visited links (which per your description looks like it), your behavior can be prone to timing attacks.

At least Gecko cascades and selector-matches both visited and unvisited rules _regardless_ of whether the link was actually visited in the first place, to prevent this kind of issues. If I'm not wrong, I should be able to query history in edge adding very expensive selectors to the `:visited` rules, and observing the time difference it takes to style a link vs another.

**Comment by FremyCompany** [comment]
@emilio I just wrote a test case and... you are not wrong. If you want to collect the dollars you are due you should file a case on this (even if that is just to give them to a charity). 

I know this is archived in some mailing list somewhere but I'm still going to delete your comment on GitHub. 

Can you please send an email to me (microsoft.com, with alias FrRemy), I'll try to direct you to the right people to track this issue. Thanks for bringing this up.

**Comment by tabatkins** [comment]
Wow, we messed up the topics for this. I'll move the comment over and delete this one.

**Comment by FremyCompany** [comment]
@victoriasu So, I just want to take an opportunity to clarify my thinking.

I do believe that if we keep the spec text as it is today, we should at the very least add a note mentioning that for most properties, :link will apply regardless of visited status, and point to [DBaron's spec](https://dbaron.org/mozilla/visited-privacy). I understand @dbaron's point of view that says that, for the purpose of the selector matching, :link and :visited are mutually exclusive (as in you cannot ever have :link:visited match), I just think that for an author reading the spec, it is important to understand that :link will, in practice, match all links except for a few select properties.



To return to the interop issue highlighted in the first post of this thread:

Globally, how it works in Edge is we kinda have a "color" and a "visited-color" property, and when a style rule applies "color:red" we set both of them; if the style came from a :visited rule however, we only set "visited-color" (we do this for a limited subset of properties, for other properties, a visited rule will not set anything when applied; converserly, when a rule contains :link we should only set the non-visited property, though whether we do this for all properties is unclear). This means that :link applies all the times for most properties, and does not apply for properties for which we got the ":link sets only non-visited field" rule right (when we do not :visited rules are still allowed to override these :link declarations that also happen to have set the visited-property, they just need higher specificity).

It looks like `background-color` is a property for which we didn't get the ":link does not set the :visited part" right, hence the interop difference in the demo above. For all other properties I tested, we behave like other browsers, so it is just a bug on us to fix this issue.

To conclude, I think as a result of this, only an editorial change should be necessary: adding a note to the spec would be great, if only because the current text caused enough confusion for browser engineers to file this issue in the first place.

**Comment by tabatkins** [comment]
Okay, I adapted @dbaron’s spec into actual spec text in PR #2105.

**Comment by byung-woo** [comment]
I recently started prototyping ':has()' pseudo class in Chrome, and there was [a mailing thread](https://groups.google.com/a/chromium.org/g/blink-dev/c/hqkcKdDrhXE/m/i2ADPY78AQAJ) about the interaction of ':has()' and ':visited' in terms of privacy issues.
 
Although the ':visited' discussion seems to be going on in another thread, and the ':has()' prototyping is very early stage, I thought it might be worth sharing the ':has()' related approach here because it is based on the current chrome ':visited' privacy approach.

The current approach to privacy considerations for ':visited' in the ':has()' argument selector is:
 - The ':visited' selector does not match if it is inside the ':has()' argument selector to prevent leaking visitedness to the link's ancestors by ':has()'. So if a ':has()' argument selector requires a matching ':visited', the style rule will not be applied.

**Comment by emilio** [comment]
@byung-woo Why do the existing mitigations not work? On Firefox, we match the link as both visited and unvisited. I don't think `:has()` introduces any new risk in that regard?

**Comment by byung-woo** [comment]
The case at above is the :visited in the :has() argument.

For example, when we have :has(:visited) {...}, the subject elements of the rule are the ancestors of the visited link element. In this case, the rule should not be applied because the ancestor elements are not link. The ':visited' in ':visited .a { ... }' will match, but the ':visited' in '.a:has(:visited)' will not match to prevent leaking visitedness out of the link elements.

It can be said that this approach follows the current chrome approach which only allows the 'descendant combinator' or 'child combinator' for the ':visited' match, because the meaning of ':has()' is technically 'ancestor combinator' or 'preceding sibling combinator'

**Comment by svgeesus** [comment]
Closed by https://github.com/w3c/csswg-drafts/pull/2105
